### PR TITLE
secboot, o/fdestate/backend: reseal with DBX update payload

### PR DESCRIPTION
**Pull request by bboozzoo**
A helper in fdestate backend to reseal with DBX update payload.

Cherry picked from #14615 

Related: SNAPDENG-32509 SNAPDENG-34135

### CODE REVIEW

File: overlord/fdestate/backend/reseal.go [hunk 1]
```diff
@@ -62,6 +62,49 @@ type StateUpdater func(role string, containerRole string, bootModes []string, mo
 
 // ResealKeyForBootChains reseals disk encryption keys with the given bootchains.
 func ResealKeyForBootChains(updateState StateUpdater, method device.SealingMethod, rootdir string, params *boot.ResealKeyForBootChainsParams, expectReseal bool) error {
+	return resealKeyForBootChains(updateState, method, rootdir,
```
Comment (valentindavid) [inline overlord/fdestate/backend/reseal.go:65]: Should we change more than one letter in the name?

**Review by valentindavid — APPROVED** [review]

File: overlord/fdestate/backend/reseal.go (hunk 1 above)
Comment (bboozzoo) [reply overlord/fdestate/backend/reseal.go]: renamed that internal function to simply `resealKeys`

File: overlord/fdestate/backend/reseal.go [hunk 2]
```diff
@@ -62,6 +62,49 @@ type StateUpdater func(role string, containerRole string, bootModes []string, mo
...
+	params *boot.ResealKeyForBootChainsParams, dbUpdate []byte,
+) error {
+	return resealKeys(updateState, method, rootdir,
+		resealInputs{
+			bootChains:        params,
+			signatureDBUpdate: dbUpdate,
+		},
+		resealOptions{
+			ExpectReseal: true,
+			Force:        true,
```
Comment (ZeyadYasser) [inline overlord/fdestate/backend/reseal.go:87]: Might be a good idea to add a small comment on why we need to force a reseal

File: overlord/fdestate/backend/reseal.go [hunk 3]
```diff
@@ -139,11 +189,22 @@ func resealRunObjectKeys(updateState StateUpdater, pbc boot.PredictableBootChain
 		return fmt.Errorf("at least one set of model-specific parameters is required")
 	}
 
+	if len(sigDBUpdate) > 0 {
+		logger.Debug("attaching DB update payload")
+		attachSignatureDBUPdate(modelParams, sigDBUpdate)
+	}
+
```
Comment (ZeyadYasser) [inline overlord/fdestate/backend/reseal.go:196]: Would it make sense to pull this up into resealKeys and drop the force reseal?

**Review by ZeyadYasser — APPROVED** [review]
LGTM, Thank you!

File: overlord/fdestate/backend/reseal.go (hunk 2 above)
Comment (bboozzoo) [reply overlord/fdestate/backend/reseal.go:87]: added

File: overlord/fdestate/backend/reseal.go (hunk 3 above)
Comment (bboozzoo) [reply overlord/fdestate/backend/reseal.go:196]: no, reseal is done at 2 points, a pre-update step, where we pass the content which will be written to the EFI variable, and then a post-update step, where the reseal will effectively observe the content of the variable

File: overlord/fdestate/backend/reseal_test.go [hunk 4]
```diff
@@ -1620,3 +1621,150 @@ func (s *resealTestSuite) TestResealKeyForBootchainsFallbackCmdline(c *C) {
...
+
+		return nil
+	})
+	defer restore()
+
+	updateCalls := 0
+	up := func(role string, containerRole string, bootModes []string, models []secboot.ModelForSealing, tpmPCRProfile []byte) error {
+		updateCalls++
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/fdestate/backend/reseal_test.go:1761]: my bad, it should be checked by the test

**Comment by bboozzoo** [comment]
Rebased and addressed the comments.

File: secboot/secboot.go [hunk 5]
```diff
@@ -97,6 +98,9 @@ type SealKeyModelParams struct {
 	EFILoadChains []*LoadChain
 	// The kernel command line
 	KernelCmdlines []string
+	// TODO move this somewhere else?
+	// The content of an update to EFI DBX
+	EFIForbiddenKeySignatureDBUpdate []byte
```
Comment (pedronis) [inline secboot/secboot.go:103]: why not call it EFISignatureDbxUpdate ?

File: secboot/secboot.go [hunk 6]
```diff
@@ -89,6 +89,7 @@ type ModelForSealing interface {
 	SignKeyID() string
 }
 
+// TODO reame to SealKeyParams?
```
Comment (pedronis) [inline secboot/secboot.go:92]: typo, also can you explain why we should rename it?

File: secboot/secboot_tpm.go [hunk 7]
```diff
@@ -510,6 +519,7 @@ func buildPCRProtectionProfile(modelParams []*SealKeyModelParams) (*sb_tpm2.PCRP
 			loadSequences,
 			sb_efi.WithSecureBootPolicyProfile(),
 			sb_efi.WithBootManagerCodeProfile(),
+			sb_efi.WithSignatureDBUpdates(updateDB...),
```
Comment (pedronis) [inline secboot/secboot_tpm.go:522]: doesn't this bit needs a test?

File: overlord/fdestate/backend/reseal.go [hunk 8]
```diff
@@ -283,3 +364,13 @@ func resealFallbackObjectKeys(updateState StateUpdater, pbc boot.PredictableBoot
 
 	return nil
 }
+
+func attachSignatureDBUPdate(params []*secboot.SealKeyModelParams, update []byte) {
```
Comment (pedronis) [inline overlord/fdestate/backend/reseal.go:368]: shouldn't all the DB in the names here be Dbx actually? as they end up in Dbx?

**Review by pedronis — COMMENTED** [review]
naming questions mostly

File: secboot/secboot.go (hunk 5 above)
Comment (bboozzoo) [reply secboot/secboot.go:103]: renamed

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**